spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HealthCheckServiceInstanceListSupplier.java

```diff
@@ -74,32 +77,33 @@ public HealthCheckServiceInstanceListSupplier(ServiceInstanceListSupplier delega
 	}
 
 	private void initInstances() {
-		delegate.get().subscribe(delegateInstances -> {
-			instances.clear();
-			instances.addAll(delegateInstances);
-		});
-
-		Flux<List<ServiceInstance>> healthCheckFlux = healthCheckFlux();
-
-		healthCheckFlux.subscribe(verifiedInstances -> {
-			healthyInstances.clear();
-			healthyInstances.addAll(verifiedInstances);
-		});
+		healthCheckDisposable = delegate.get().doOnNext(delegateInstances -> {
+			instances = Collections.unmodifiableList(new ArrayList<>(delegateInstances));
+		})
+		.thenMany(healthCheckFlux()).subscribeOn(Schedulers.parallel())
+		.subscribe(verifiedInstances -> healthyInstances = verifiedInstances);
 	}
 
 	protected Flux<List<ServiceInstance>> healthCheckFlux() {
-		return Flux.create(emitter -> Schedulers
-				.newSingle("Health Check Verifier: " + getServiceId(), true)
-				.schedulePeriodically(() -> {
-					List<ServiceInstance> verifiedInstances = new ArrayList<>();
-					Flux.fromIterable(instances).filterWhen(this::isAlive)
-							.subscribe(serviceInstance -> {
-								verifiedInstances.add(serviceInstance);
-								emitter.next(verifiedInstances);
-							});
-				}, healthCheck.getInitialDelay(), healthCheck.getInterval().toMillis(),
-						TimeUnit.MILLISECONDS),
-				FluxSink.OverflowStrategy.LATEST);
+		return Flux.defer(() -> {
+			List<ServiceInstance> result = new CopyOnWriteArrayList<>();
+
+			List<Mono<ServiceInstance>> checks = new ArrayList<>();
+			for (ServiceInstance instance : instances) {
+				Mono<ServiceInstance> alive = isAlive(instance)
+						.onErrorResume(throwable -> Mono.empty())
```

Comment:
should some logs be added here?  
if so `debug` or `warn`?

Reply:
yes, debug is my first instinct.